server/src/main/java/org/opensearch/search/DefaultSearchContext.java

```diff
@@ -902,11 +921,61 @@ public Profilers getProfilers() {
     @Override
     public boolean shouldUseConcurrentSearch() {
         assert requestShouldUseConcurrentSearch.get() != null : "requestShouldUseConcurrentSearch must be set";
-        return concurrentSearchSettingsEnabled && Boolean.TRUE.equals(requestShouldUseConcurrentSearch.get());
+        assert concurrentSearchMode != null : "concurrentSearchMode must be set";
+        return (concurrentSearchMode.equals(CONCURRENT_SEGMENT_SEARCH_MODE_AUTO)
+            || concurrentSearchMode.equals(CONCURRENT_SEGMENT_SEARCH_MODE_ALL))
+            && Boolean.TRUE.equals(requestShouldUseConcurrentSearch.get());
+    }
+
+    private boolean evaluateAutoMode() {
+
+        // filter out deciders that want to opt-out of decision-making
+        final Set<ConcurrentSearchDecider> filteredDeciders = concurrentSearchDeciders.stream()
+            .filter(concurrentSearchDecider -> concurrentSearchDecider.canEvaluateForIndex(indexService.getIndexSettings()))
+            .collect(Collectors.toSet());
+        // evaluate based on concurrent search query visitor
+        if (filteredDeciders.size() > 0) {
+            ConcurrentSearchVisitor concurrentSearchVisitor = new ConcurrentSearchVisitor(
+                filteredDeciders,
+                indexService.getIndexSettings()
+            );
+            if (request().source() != null && request().source().query() != null) {
+                QueryBuilder queryBuilder = request().source().query();
+                queryBuilder.visit(concurrentSearchVisitor);
+            }
+        }
+        final List<ConcurrentSearchDecision> decisions = filteredDeciders.stream()
+            .map(ConcurrentSearchDecider::getConcurrentSearchDecision)
+            .filter(Objects::nonNull)
+            .collect(Collectors.toList());
```

Comment:
Can we unroll above to a traditional loop and dump the returned decisions in debug log along with collecting it ?